{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Library/Python/2.7/site-packages/pandas/rpy/__init__.py:8: FutureWarning: The pandas.rpy module is deprecated and will be removed in a future version. We refer to external packages like rpy2, found here: http://rpy.sourceforge.net\n",
      "  \"like rpy2, found here: http://rpy.sourceforge.net\", FutureWarning)\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import gzip\n",
    "import math\n",
    "import pandas\n",
    "import StringIO\n",
    "import textwrap\n",
    "import dendropy\n",
    "from ngs_parsers import VCF\n",
    "import pandas.rpy.common as com\n",
    "import rpy2.robjects as robjects\n",
    "from itertools import combinations\n",
    "from collections import OrderedDict\n",
    "from rpy2.robjects.packages import importr\n",
    "import numpy as np\n",
    "\n",
    "vcf = VCF.VCF('/Users/testudines/DATA/5M/VCF/RNAseq_sample_lab_ids_v2.vcf.gz')\n",
    "vcf.populations  =   {'AGAW': ['ETAI014','ETAI015','ETAI018',\n",
    "                               'ETAI020','ETAI021','ETAI023',\n",
    "                               'ETAI026','ETAI028','ETAI029',\n",
    "                               'ETAI030','ETAI032','ETAI033',\n",
    "                               'ETAS003','ETAS024','ETAS026',\n",
    "                               'ETAS030','ETAS031','ETAS034',\n",
    "                               'ETAS035'],\n",
    "                    'AMHARA': ['ETAM041','ETAM042','ETAM043',\n",
    "                               'ETAM044','ETAM045','ETAM046',\n",
    "                               'ETAM047','ETAM048','ETAM049',\n",
    "                               'ETAM051','ETAM052','ETAM053',\n",
    "                               'ETAM055','ETAM058','ETAM060',\n",
    "                               'ETAM061','ETAM063','ETAM070',\n",
    "                               'ETAM080'],\n",
    "                   'ARGOBBA': ['ETAG001','ETAG002','ETAG004',\n",
    "                               'ETAG007','ETAG008','ETAG009',\n",
    "                               'ETAG013','ETAG014','ETAG015',\n",
    "                               'ETAG017','ETAG018','ETAG019',\n",
    "                               'ETAG022','ETAG023','ETAG027',\n",
    "                               'ETAG029'],\n",
    "                      'DIZI': ['ETDI001','ETDI002','ETDI006',\n",
    "                               'ETDI008','ETDI011','ETDI013',\n",
    "                               'ETDI014','ETDI015','ETDI019',\n",
    "                               'ETDI020','ETDI021','ETDI022',\n",
    "                               'ETDI023','ETDI027','ETDI028',\n",
    "                               'ETDI036','ETDI037','ETDI039',\n",
    "                               'ETDI040'],\n",
    "                   'HADZABE': ['TZHZ401','TZHZ402','TZHZ442',\n",
    "                               'TZHZ446','TZHZ457','TZHZ463',\n",
    "                               'TZHZ467','TZHZ469','TZHZ476',\n",
    "                               'TZHZ479','TZHZ488','TZHZ492',\n",
    "                               'TZHZ497','TZHZ505','TZHZ506',\n",
    "                               'TZHZ509','TZHZ533','TZHZ550',\n",
    "                               'TZHZ555'],\n",
    "                     'MURSI': ['ETMS002','ETMS004','ETMS006',\n",
    "                               'ETMS008','ETMS009','ETMS011',\n",
    "                               'ETMS012','ETMS013','ETMS014',\n",
    "                               'ETMS018','ETMS020','ETMS021',\n",
    "                               'ETMS022','ETMS023','ETMS027',\n",
    "                               'ETMS029','ETMS033','ETMS034',\n",
    "                               'ETMS036'],\n",
    "                     'SABUE': ['ETSB001','ETSB002','ETSB004',\n",
    "                               'ETSB005','ETSB006','ETSB008',\n",
    "                               'ETSB009','ETSB012','ETSB013',\n",
    "                               'ETSB014','ETSB016','ETSB023',\n",
    "                               'ETSB027','ETSB029','ETSB031',\n",
    "                               'ETSB035','ETSB036','ETSB039'],\n",
    "                   'SANDAWE': ['TZSW140','TZSW142','TZSW145',\n",
    "                               'TZSW156','TZSW161','TZSW162',\n",
    "                               'TZSW174','TZSW177','TZSW180',\n",
    "                               'TZSW182','TZSW188','TZSW193',\n",
    "                               'TZSW197','TZSW205',\n",
    "                               'TZSW209'],\n",
    "                     'WEYTO': ['ETWE002','ETWE005','ETWE007',\n",
    "                               'ETWE008','ETWE009','ETWE010',\n",
    "                               'ETWE011','ETWE014','ETWE016',\n",
    "                               'ETWE017','ETWE020','ETWE023',\n",
    "                               'ETWE027','ETWE030','ETWE032',\n",
    "                               'ETWE033','ETWE036','ETWE038']}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.panel.Panel'>\n",
      "Dimensions: 26 (items) x 4 (major_axis) x 9 (minor_axis)\n",
      "Items axis: 0 to 25\n",
      "Major_axis axis: 0 to 3\n",
      "Minor_axis axis: AGAW to WEYTO\n",
      "(DIZI:-0.0002468941188,(((AGAW:0.008242019559,ARGOBBA:-0.004578337855):0.003652023069,AMHARA:0.002076544695):0.01621587507,WEYTO:-0.005916750952):0.02473709469,((SABUE:0.002250552152,SANDAWE:-0.000655500069):0.005910848068,(MURSI:0.0153816623,HADZABE:0.03778400337):0.009288255912):0.005116271668);\n"
     ]
    }
   ],
   "source": [
    "# def Wrights_Fst(ac, pop_ids):\n",
    "#     \"\"\"Does not correct for sample size.\"\"\"\n",
    "    \n",
    "#     af = ac.ix[:,0].T[pop_ids] / ac.sum(axis=1).T[pop_ids]\n",
    "#     af_dif = af.subtract(af.mean(axis=1), axis=0)\n",
    "#     s2 = np.sum(af_dif.pow(2) / (len(pop_ids) -1), axis=1)\n",
    "#     pm = af.mean(axis=1)\n",
    "#     Fst = s2 / (pm * (1 - pm))\n",
    "#     Fst = Fst.mean()\n",
    "#     return Fst\n",
    "\n",
    "\n",
    "# def Weirs_Fst(ac, h, ps, pop_ids):\n",
    "    \n",
    "#     n = ps[pop_ids].mean(axis=1)\n",
    "#     r = float(len(pop_ids))\n",
    "#     hm = h[pop_ids].mean(axis=1)\n",
    "\n",
    "#     af = ac.ix[:,0].T[pop_ids] / ac.sum(axis=1).T[pop_ids] # calc allele freq (p)\n",
    "#     pm = af.mean(axis=1)                                   # calc. p bar\n",
    "#     af_dif = af.subtract(pm, axis=0)\n",
    "#     s2 = np.sum(af_dif.pow(2) / (len(pop_ids) -1), axis=1)\n",
    "\n",
    "\n",
    "#     N = s2 - (1 / (2*n-1)) * ((pm*(1 - pm))-(((r - 1)/r) * s2)-(hm/4))\n",
    "#     D = (pm*(1 - pm)) + s2/r\n",
    "#     Fst = N.sum() / D.sum()\n",
    "#     return Fst\n",
    "\n",
    "\n",
    "# # def subset_pops(dct, pops):\n",
    "# #     return OrderedDict([ i for i in dct.iteritems() if i[0] in pops])\n",
    "    \n",
    "\n",
    "# def make_empty_df(pops):\n",
    "#     data = np.zeros((len(pops),len(pops)))\n",
    "#     df = pandas.DataFrame(data, index=pops, columns=pops)\n",
    "#     return df\n",
    "    \n",
    "\n",
    "# def create_tables(vcf):\n",
    "\n",
    "#     allele_counts = {}\n",
    "#     hets = []\n",
    "#     pop_sizes = []\n",
    "#     snp_ids = []\n",
    "\n",
    "#     for c, i in enumerate(vcf.vcf_file_iterator()):\n",
    "#         if c > 25: break\n",
    "        \n",
    "#         ac = vcf.calc_allele_counts(i)\n",
    "\n",
    "#         h  = vcf.calc_heterozygosity(i)\n",
    "#         ps = vcf.get_population_sizes(i)\n",
    "\n",
    "#         # snp_id = '{}_{}'.format(i['CHROM'],i['POS'])\n",
    "#         snp_id = c\n",
    "#         snp_ids.append(snp_id)\n",
    "#         hets.append(h)\n",
    "#         pop_sizes.append(ps)\n",
    "\n",
    "#         allele_counts[snp_id] = pandas.DataFrame(ac)\n",
    "    \n",
    "    \n",
    "#     ac = pandas.Panel(allele_counts)\n",
    "#     h = pandas.DataFrame(hets, index=snp_ids)\n",
    "#     ps = pandas.DataFrame(pop_sizes, index=snp_ids)\n",
    "    \n",
    "#     return ac, h, ps\n",
    "\n",
    "# def mean_pairwise_Fst(pops, min_sample_count=7):\n",
    "    \n",
    "#     # create empty distance matrix\n",
    "#     df = make_empty_df(pops)\n",
    "    \n",
    "#     ac, h, ps = create_tables(vcf)\n",
    "#     print ac\n",
    "#     for pop1, pop2 in combinations(pops,2):\n",
    "        \n",
    "#         Fst = Weirs_Fst(ac, h, ps, [pop1, pop2])\n",
    "#         df.ix[pop1,pop2] = Fst\n",
    "#         df.ix[pop2,pop1] = Fst\n",
    "\n",
    "#     return df\n",
    "\n",
    "        \n",
    "# pops = vcf.populations.keys()\n",
    "# df = mean_pairwise_Fst(pops)\n",
    "# tree = calc_nj_tree(df)\n",
    "# print tree\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "allele counts done.\n",
      "(AGAW:0.007493583098,AMHARA:0.007838946313,((((((MURSI:0.01208370322,SABUE:0.04218702663):0.005896015748,HADZABE:0.04393485503):0.002859667951,SANDAWE:0.01477960955):0.007076610095,DIZI:0.01301411754):0.01362818158,WEYTO:0.021065968):0.003648512114,ARGOBBA:0.008202890016):0.0003616729256);\n"
     ]
    }
   ],
   "source": [
    "import StringIO \n",
    "from dendropy import Tree\n",
    "\n",
    "def make_empty_df(pops):\n",
    "    data = np.zeros((len(pops),len(pops)))\n",
    "    df = pandas.DataFrame(data, index=pops, columns=pops)\n",
    "    return df\n",
    "\n",
    "def extract_allele_counts(vcf):\n",
    "    \n",
    "    pops = []\n",
    "    samples = []\n",
    "    for p, spls in vcf.populations.iteritems():\n",
    "        for s in spls:\n",
    "            pops.append(p)\n",
    "            samples.append(s)\n",
    "    \n",
    "    df =[]\n",
    "    for c, i in enumerate(vcf.vcf_file_iterator()):\n",
    "        #if c > 500: break\n",
    "            \n",
    "        ac = {}   \n",
    "        for s in samples:\n",
    "\n",
    "            if i[s] is None:\n",
    "                ac[s] = None\n",
    "\n",
    "            elif i[s]['GT'] == '0/0':\n",
    "                ac[s] = 0\n",
    "\n",
    "            elif i[s]['GT'] == '1/0' or i[s]['GT'] == '0/1':\n",
    "                ac[s] = 1\n",
    "\n",
    "            elif i[s]['GT'] == '1/1':\n",
    "                ac[s] = 2\n",
    "\n",
    "        df.append(ac)\n",
    "\n",
    "    df = pandas.DataFrame(df)\n",
    "    return df\n",
    "\n",
    "\n",
    "def ac_2_af(df, pops):\n",
    "    \"\"\"Convert table of allele counts to table of allele frequencies.\"\"\"  \n",
    "    # Calculate Allele Frequencies  \n",
    "    new_df = {pop: df[pops[pop]].sum(axis=1).divide(df[pops[pop]].shape[1]*2.0) for pop in pops}\n",
    "    return pandas.DataFrame(new_df)\n",
    "\n",
    "\n",
    "def Wrights_Fst(df, ps_dict, bootstrap=False):\n",
    "    \"\"\"Does not correct for sample size.\"\"\"\n",
    "    pop_ids = ps_dict.keys()\n",
    "    af = ac_2_af(df, ps_dict, bootstrap)\n",
    "    p = af.ix[:,pop_ids]\n",
    "    p_bar = af.ix[:,pop_ids].mean(axis=1)\n",
    "    p_dif = p.subtract(p_bar, axis=0)\n",
    "    s2 = np.sum(p_dif.pow(2) / (len(pop_ids) -1), axis=1)\n",
    "    Fst = s2 / (p_bar * (1 - p_bar))\n",
    "    Fst = Fst.mean()\n",
    "    if np.isinf(Fst) == True:\n",
    "        Fst = 0\n",
    "    return Fst\n",
    "\n",
    "\n",
    "def Weirs_Fst(af, h, ps, ps_dict, bootstrap=False):\n",
    "    \n",
    "    pop_ids = ps_dict.keys()\n",
    "    \n",
    "    r = float(len(pop_ids))\n",
    "    n_bar = ps[pop_ids].mean(axis=1)\n",
    "    h_bar = h[pop_ids].mean(axis=1)\n",
    "    p_bar = af.mean(axis=1)                                   # calc. p bar    \n",
    "    \n",
    "    af_dif = af.subtract(p_bar, axis=0)\n",
    "    s2 = np.sum(af_dif.pow(2) / (len(pop_ids) -1), axis=1)\n",
    "\n",
    "    N = s2 - (1 / (2*n_bar-1)) * (p_bar*(1 - p_bar)-(r - 1)/r * s2 - h_bar/4)\n",
    "    D = (p_bar*(1 - p_bar)) + s2/r\n",
    "    Fst = N.sum() / D.sum()\n",
    "    return Fst\n",
    "\n",
    "\n",
    "def ac_2_het(df, pops):\n",
    "    ones = df==1\n",
    "    new_df = {pop : ones[pops[pop]].sum(axis=1).divide(ones[pops[pop]].shape[1]) for pop in pops}\n",
    "    return pandas.DataFrame(new_df)\n",
    "\n",
    "\n",
    "def ac_2_ps(df, pops):\n",
    "    new_df = {pop : df[pops[pop]].count(axis=1) for pop in pops}\n",
    "    return pandas.DataFrame(new_df)   \n",
    "\n",
    "\n",
    "def mean_pairwise_Fst(ac, pops):\n",
    "    \n",
    "    #pops = vcf.populations\n",
    "    \n",
    "    # create empty distance matrix\n",
    "    final_df = make_empty_df(pops)\n",
    "    \n",
    "    h = ac_2_het(ac, pops)\n",
    "    ps = ac_2_ps(ac, pops)\n",
    "    \n",
    "    for pop1, pop2 in combinations(pops,2):\n",
    "        \n",
    "        ps_dict = {k: pops[k] for k in (pop1, pop2)}\n",
    "        af = ac_2_af(ac, ps_dict)\n",
    "        \n",
    "        Fst = Weirs_Fst(af, h, ps, ps_dict)\n",
    "        \n",
    "        final_df.ix[pop1,pop2] = Fst\n",
    "        final_df.ix[pop2,pop1] = Fst\n",
    "\n",
    "    return (final_df, h, ps, af)\n",
    "\n",
    "\n",
    "def make_boot_rep(pops):\n",
    "    # Bootstraping\n",
    "    #   Resample with replacement each population.\n",
    "    \n",
    "    return {pop: list(np.random.choice(samples, len(samples), replace=True)) \\\n",
    "            for pop, samples in pops.iteritems()}\n",
    "\n",
    "\n",
    "def calc_nj_tree(mtx):\n",
    "    mtx = com.convert_to_r_matrix(mtx)\n",
    "    ape = importr('ape', on_conflict='warn')\n",
    "    as_dist = robjects.r['as.dist']\n",
    "    NJ = robjects.r['nj']\n",
    "    tree = NJ(as_dist(mtx))\n",
    "    #pt = robjects.r['plot']\n",
    "    #pt(tree)\n",
    "    nwk = ape.write_tree(tree)[0]\n",
    "    return nwk\n",
    "    \n",
    "pops = vcf.populations\n",
    "ac = extract_allele_counts(vcf)\n",
    "print 'allele counts done.'\n",
    "final_df, h, ps, af =  mean_pairwise_Fst(ac, pops)\n",
    "tree = calc_nj_tree(final_df)\n",
    "print tree\n",
    "\n",
    "# ntrees = []\n",
    "# trees= []\n",
    "# for i in range(0,500):\n",
    "\n",
    "#     if i % 25 == 0:\n",
    "#         print i\n",
    "#     b_pops = make_boot_rep(pops)\n",
    "#     #print b_pops\n",
    "#     final_df, h, ps, af =  mean_pairwise_Fst(ac, b_pops)\n",
    "#     tree = calc_nj_tree(final_df)\n",
    "#     ntrees.append(tree)\n",
    "#     tree = Tree.get_from_string(tree, schema=\"newick\")\n",
    "#     trees.append(tree)\n",
    "\n",
    "# trees = dendropy.TreeList(trees)\n",
    "# con_tree = trees.consensus(min_freq=0.50)\n",
    "# con_tree.as_string('newick')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MURSI</th>\n",
       "      <th>AMHARA</th>\n",
       "      <th>AGAW</th>\n",
       "      <th>SABUE</th>\n",
       "      <th>HADZABE</th>\n",
       "      <th>WEYTO</th>\n",
       "      <th>SANDAWE</th>\n",
       "      <th>DIZI</th>\n",
       "      <th>ARGOBBA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MURSI</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.054183</td>\n",
       "      <td>0.053368</td>\n",
       "      <td>0.054271</td>\n",
       "      <td>0.062895</td>\n",
       "      <td>0.061420</td>\n",
       "      <td>0.039172</td>\n",
       "      <td>0.038306</td>\n",
       "      <td>0.051566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AMHARA</th>\n",
       "      <td>0.054183</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.015333</td>\n",
       "      <td>0.086608</td>\n",
       "      <td>0.081127</td>\n",
       "      <td>0.032534</td>\n",
       "      <td>0.046172</td>\n",
       "      <td>0.038821</td>\n",
       "      <td>0.016424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGAW</th>\n",
       "      <td>0.053368</td>\n",
       "      <td>0.015333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.086233</td>\n",
       "      <td>0.080721</td>\n",
       "      <td>0.032102</td>\n",
       "      <td>0.045626</td>\n",
       "      <td>0.038988</td>\n",
       "      <td>0.016038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SABUE</th>\n",
       "      <td>0.054271</td>\n",
       "      <td>0.086608</td>\n",
       "      <td>0.086233</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.091037</td>\n",
       "      <td>0.093310</td>\n",
       "      <td>0.070510</td>\n",
       "      <td>0.059145</td>\n",
       "      <td>0.084790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HADZABE</th>\n",
       "      <td>0.062895</td>\n",
       "      <td>0.081127</td>\n",
       "      <td>0.080721</td>\n",
       "      <td>0.091037</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.090051</td>\n",
       "      <td>0.057404</td>\n",
       "      <td>0.065630</td>\n",
       "      <td>0.079795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WEYTO</th>\n",
       "      <td>0.061420</td>\n",
       "      <td>0.032534</td>\n",
       "      <td>0.032102</td>\n",
       "      <td>0.093310</td>\n",
       "      <td>0.090051</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.056742</td>\n",
       "      <td>0.047315</td>\n",
       "      <td>0.033766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SANDAWE</th>\n",
       "      <td>0.039172</td>\n",
       "      <td>0.046172</td>\n",
       "      <td>0.045626</td>\n",
       "      <td>0.070510</td>\n",
       "      <td>0.057404</td>\n",
       "      <td>0.056742</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.039126</td>\n",
       "      <td>0.045412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DIZI</th>\n",
       "      <td>0.038306</td>\n",
       "      <td>0.038821</td>\n",
       "      <td>0.038988</td>\n",
       "      <td>0.059145</td>\n",
       "      <td>0.065630</td>\n",
       "      <td>0.047315</td>\n",
       "      <td>0.039126</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.037714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARGOBBA</th>\n",
       "      <td>0.051566</td>\n",
       "      <td>0.016424</td>\n",
       "      <td>0.016038</td>\n",
       "      <td>0.084790</td>\n",
       "      <td>0.079795</td>\n",
       "      <td>0.033766</td>\n",
       "      <td>0.045412</td>\n",
       "      <td>0.037714</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            MURSI    AMHARA      AGAW     SABUE   HADZABE     WEYTO   SANDAWE  \\\n",
       "MURSI    0.000000  0.054183  0.053368  0.054271  0.062895  0.061420  0.039172   \n",
       "AMHARA   0.054183  0.000000  0.015333  0.086608  0.081127  0.032534  0.046172   \n",
       "AGAW     0.053368  0.015333  0.000000  0.086233  0.080721  0.032102  0.045626   \n",
       "SABUE    0.054271  0.086608  0.086233  0.000000  0.091037  0.093310  0.070510   \n",
       "HADZABE  0.062895  0.081127  0.080721  0.091037  0.000000  0.090051  0.057404   \n",
       "WEYTO    0.061420  0.032534  0.032102  0.093310  0.090051  0.000000  0.056742   \n",
       "SANDAWE  0.039172  0.046172  0.045626  0.070510  0.057404  0.056742  0.000000   \n",
       "DIZI     0.038306  0.038821  0.038988  0.059145  0.065630  0.047315  0.039126   \n",
       "ARGOBBA  0.051566  0.016424  0.016038  0.084790  0.079795  0.033766  0.045412   \n",
       "\n",
       "             DIZI   ARGOBBA  \n",
       "MURSI    0.038306  0.051566  \n",
       "AMHARA   0.038821  0.016424  \n",
       "AGAW     0.038988  0.016038  \n",
       "SABUE    0.059145  0.084790  \n",
       "HADZABE  0.065630  0.079795  \n",
       "WEYTO    0.047315  0.033766  \n",
       "SANDAWE  0.039126  0.045412  \n",
       "DIZI     0.000000  0.037714  \n",
       "ARGOBBA  0.037714  0.000000  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'[&U] (ARGOBBA,(WEYTO:0.0343027933097,(DIZI:0.0255149399844,(SANDAWE:0.0301677678063,(HADZABE:0.0558620214162,(SABUE:0.0547466835168,MURSI:0.0249934620731)1.0000:0.00601069306252)0.8380:0.00287351697484)1.0000:0.00667493249322)1.0000:0.0134846993311)0.9960:0.00349287435667,(AMHARA:0.0204171210698,AGAW:0.020374795018)0.5260:0.00089361577623);\\n'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "con_tree = trees.consensus(min_freq=0.20)\n",
    "con_tree.as_string('newick')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "mtx = com.convert_to_r_matrix(final_df)\n",
    "ape = importr('ape', on_conflict='warn')\n",
    "as_dist = robjects.r['as.dist']\n",
    "NJ = robjects.r['nj']\n",
    "tree = NJ(as_dist(mtx)) #,\"unrooted\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for t in ntrees:\n",
    "    #print t\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-9-2c7141916938>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mac\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mextract_allele_counts\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvcf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mfinal_df\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mh\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mps\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maf\u001b[0m \u001b[0;34m=\u001b[0m  \u001b[0mmean_pairwise_Fst\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mac\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpops\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mtree\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcalc_nj_tree\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfinal_df\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mtree\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-6-a6d041226f9b>\u001b[0m in \u001b[0;36mextract_allele_counts\u001b[0;34m(vcf)\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m     \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 19\u001b[0;31m     \u001b[0;32mfor\u001b[0m \u001b[0mc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0menumerate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvcf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvcf_file_iterator\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     20\u001b[0m         \u001b[0;31m#if c > 500: break\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     21\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/testudines/Code/pypgen2/ngs_parsers/VCF.pyc\u001b[0m in \u001b[0;36mvcf_file_iterator\u001b[0;34m(self, as_dict)\u001b[0m\n\u001b[1;32m    223\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstartswith\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"#\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    224\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mas_dict\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 225\u001b[0;31m                     \u001b[0;32myield\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparse_vcf_line\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mline\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mempty_vcf_line\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    226\u001b[0m                 \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    227\u001b[0m                     \u001b[0;32myield\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/testudines/Code/pypgen2/ngs_parsers/VCF.pyc\u001b[0m in \u001b[0;36mparse_vcf_line\u001b[0;34m(self, pos, vcf_line_dict)\u001b[0m\n\u001b[1;32m    273\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    274\u001b[0m                 \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 275\u001b[0;31m                     \u001b[0mgenotype\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mzip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msample_format\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgenotype\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\":\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    276\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    277\u001b[0m                     \u001b[0;31m# CONVERT STRINGS TO APPOPRIATE TYPES (INTS, FLOATS, ETC.)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "ac = extract_allele_counts(vcf)\n",
    "final_df, h, ps, af =  mean_pairwise_Fst(ac, pops)\n",
    "tree = calc_nj_tree(final_df)\n",
    "tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
